[Excel] Refresh Excel low-engagement articles (#5779)
OfficeDev/office-js-docs-pr: docs/excel/excel-data-types-concepts.md

---
title: Use Excel JavaScript API data types
description: Learn when to use `valuesAsJson`, formatted numbers, entity values, linked entities, web images, and enhanced errors in Excel add-ins.
ms.date: 06/03/2026
ms.topic: concept-article
ms.custom: scenarios:getting-started
ai-usage: ai-assisted
ms.localizationpriority: high
---

# Use data types in Excel add-ins

> When your add-in needs more than strings, numbers, and booleans, use Excel data types. Data types let you return rich values such as formatted dates, entity cards, linked records, and web images while still supporting worksheet calculations.

Reviewer: ```suggestion
When your add-in needs more than strings, numbers, and booleans, use Excel data types. Data types let you return enhanced values such as formatted dates, entity cards, linked records, and web images while still supporting worksheet calculations.
```

This article explains the `valuesAsJson` API that powers data types and shows when to use the main cell value types. For a feature overview, see [Overview of data types in Excel add-ins](excel-data-types-overview.md).

To try these concepts right away, open [**Script Lab**](../overview/explore-with-script-lab.md) in Excel and browse the **Data types** samples in the **Samples** library.

## The `valuesAsJson` property

The `valuesAsJson` property is the main API for reading and writing Excel data types. The singular `valueAsJson` property on [NamedItem](/javascript/api/excel/excel.nameditem) serves the same purpose for a single named item.

> `valuesAsJson` expands on properties such as [Range.values](/javascript/api/excel/excel.range#excel-excel-range-values-member). The `values` property only returns one of the four basic cell value types: string, number, boolean, or error. By contrast, `valuesAsJson` returns the richer JSON structure for those basic types and for data types such as formatted numbers, entities, and web images.

Reviewer: Per style guide and global English guidelines we should prioritize plain, literal language 
```suggestion
`valuesAsJson` expands on properties such as [Range.values](/javascript/api/excel/excel.range#excel-excel-range-values-member). The `values` property only returns one of the four basic cell value types: string, number, boolean, or error. By contrast, `valuesAsJson` returns an expanded JSON structure for those basic types and for data types such as formatted numbers, entities, and web images.
```

To experiment with formatted number values, open [**Script Lab**](../overview/explore-with-script-lab.md) and run the [Data types: Formatted numbers](https://github.com/OfficeDev/office-js-snippets/blob/prod/samples/excel/20-data-types/data-types-formatted-number.yaml) sample.

## Basic cell values

You can add properties to basic Excel values to associate extra information with them. This pattern works with the **string**, **double**, and **boolean** basic types. Use it when you want a simple cell value to carry related fields without turning the value into a full entity.

For example, a bill total can include related fields such as **Drinks**, **Food**, **Tax**, and **Tip**.

:::image type="content" source="../images/data-type-basic-fields.png" alt-text="Screenshot of the drinks, food, tax, and tip fields shown for the selected cell value.":::

For the complete walkthrough, see [Add properties to Excel basic cell values](excel-data-types-add-properties-to-basic-cell-values.md).

## Entity values

> Use an entity value when you want one cell to represent a rich business object. An [EntityCellValue](/javascript/api/excel/excel.entitycellvalue) can store text, nested data types, and arrays, and Excel can display that data in an entity card.

Reviewer: "Rich business object" sounds like meaningless AI-speak. 
```suggestion
An [EntityCellValue](/javascript/api/excel/excel.entitycellvalue) can store text, nested data types, and arrays, and Excel can display that data in an entity card.
```

> - If a nested entity includes a `referencedValues` array, the JavaScript Excel API throws a `GeneralException` error in add-in or script code, or Excel displays a **#VALUE!** error when a custom function produces the value.
>
> To reference values from a nested entity, use [ReferenceCellValue](/javascript/api/excel/excel.referencecellvalue) indices that point to the root entity's `referencedValues` array.

To explore entity data types, open [**Script Lab**](../overview/explore-with-script-lab.md) and run [Data types: Create entity cards from data in a table](https://github.com/OfficeDev/office-js-snippets/blob/prod/samples/excel/20-data-types/data-types-entity-values.yaml). For deeper examples, see [Data types: Entity values with references](https://github.com/OfficeDev/office-js-snippets/blob/prod/samples/excel/20-data-types/data-types-references.yaml) and [Data types: Entity value attribution properties](https://github.com/OfficeDev/office-js-snippets/blob/prod/samples/excel/20-data-types/data-types-entity-attribution.yaml).

### Linked entity cell values

> [LinkedEntityCellValue](/javascript/api/excel/excel.linkedentitycellvalue) represents an entity that's connected to an external data source. Use linked entities when you need rich cards for large or frequently updated data sets and you don't want to load all details into the workbook at once.

Reviewer: ```suggestion
[LinkedEntityCellValue](/javascript/api/excel/excel.linkedentitycellvalue) represents an entity that's connected to an external data source. Use linked entities when you need cards for large or frequently updated data sets and you don't want to load all details into the workbook at once.
```
